### PR TITLE
WIP: attrd sync point support

### PR DESCRIPTION
**Pull request by clumens**
Here is the very beginning of the sync point code around attrd.  Right now, only `--wait local` for the update operation is supported.  I haven't yet found a good place to add the global sync point.  I'm also not sure if I'm handling multiple call IDs properly.  Additionally, do we want to support sync points for all operations or just update?  It seems like it could be more broadly useful.

### CODE REVIEW

File: daemons/attrd/attrd_sync.c [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    wl->client = request->ipc_client;
+    wl->callid = strdup(crm_element_value(request->xml, F_ATTRD_CALLID));
+
+    if (pcmk__str_eq(sync_point, PCMK__ATTRD_SYNC_POINT_LOCAL, pcmk__str_none)) {
+        wl->sync_point = attrd_sync_point_local;
+    } else if (pcmk__str_eq(sync_point, PCMK__ATTRD_SYNC_POINT_ALL, pcmk__str_none)) {
+        wl->sync_point = attrd_sync_point_all;
+    } else {
+        return;
```
Comment (kgaillot) [inline daemons/attrd/attrd_sync.c:84]: free entry

File: daemons/attrd/attrd_sync.c [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        iter = client->next;
+
+        if (wl->sync_point != sync_point) {
+            continue;
+        }
+
+        crm_debug("Alerting client %s(%s) for reached %s sync point",
+                  pcmk__client_name(wl->client),
+                  wl->callid, sync_point_str(wl->sync_point));
```
Comment (kgaillot) [inline daemons/attrd/attrd_sync.c:112]: We need to alert only for the particular attribute requested

File: daemons/attrd/attrd_sync.c [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        g_list_free(client);
+        free_waitlist_node(wl);
+    }
+}
+
+bool
+attrd_client_on_waitlist(pcmk__request_t *request)
+{
+    const char *sync_point = crm_element_value(request->xml, PCMK__XA_ATTR_SYNC_POINT);
+    struct waitlist_node wl;
```
Comment (kgaillot) [inline daemons/attrd/attrd_sync.c:128]: Always a good idea to zero out bare structs with `= { 0, };`

**Review by kgaillot — COMMENTED** [review]
Good start

File: include/crm/common/attrd_internal.h [hunk 4]
```diff
@@ -14,6 +14,8 @@
 extern "C" {
 #endif
 
+#define F_ATTRD_CALLID          "attrd_callid"
```
[the remark this answers is not in the document]
Comment (clumens) [reply include/crm/common/attrd_internal.h:17]: Done.  I had a feeling there was something that could be reused here.

File: lib/common/ipc_attrd.c [hunk 5]
```diff
@@ -140,9 +142,19 @@ create_attrd_op(const char *user_name)
...
+     */
+    call_id++;
+    if (call_id < 1) {
+        call_id = 1;
+    }
+
     crm_xml_add(attrd_op, F_TYPE, T_ATTRD);
     crm_xml_add(attrd_op, F_ORIG, (crm_system_name? crm_system_name: "unknown"));
     crm_xml_add(attrd_op, PCMK__XA_ATTR_USER, user_name);
+    crm_xml_add_int(attrd_op, F_ATTRD_CALLID, call_id);
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/ipc_attrd.c:157]: Fixed.  If nothing else, seeing call-id in the XML when we don't expect it during debugging might be confusing.

File: daemons/attrd/attrd_sync.c [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "pacemaker-attrd.h"
+
+static GList *waitlist = NULL;
+
+struct waitlist_node {
+    enum attrd_sync_point sync_point;
+
+    pcmk__client_t *client;
+    char *callid;
```
[the remark this answers is not in the document]
Comment (clumens) [reply daemons/attrd/attrd_sync.c:22]: Yeah, definitely.  Fixed.

File: daemons/attrd/attrd_sync.c (hunk 1 above)
Comment (clumens) [reply daemons/attrd/attrd_sync.c:84]: Fixed.

File: daemons/attrd/attrd_sync.c (hunk 3 above)
Comment (clumens) [reply daemons/attrd/attrd_sync.c:128]: I would have expected that to either be guaranteed initialized by the C standard (it's not), or caught by some gcc flag we're already using (it's not).  I wonder if it's worth adding `-Wuninitialized`.  Fixed anyway.

File: daemons/attrd/attrd_sync.c (hunk 2 above)
Comment (clumens) [reply daemons/attrd/attrd_sync.c:112]: Good point.  This also means I need to add the attribute's name to the waitlist.  Fixed.

File: daemons/attrd/attrd_sync.c (hunk 3 above)
Comment (clumens) [reply daemons/attrd/attrd_sync.c:128]: `-Wuninitialized` wouldn't catch this.  Incidentally, the source currently builds cleanly with that option  We could add it to make sure that continues, but I'm not sold on it.

Comment (clumens) [reply daemons/attrd/attrd_sync.c:128]: Dropped this function entirely.

File: daemons/attrd/attrd_sync.c [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    if (sync_point == NULL) {
+        return;
+    }
+
+    wl = calloc(sizeof(struct waitlist_node), 1);
+
+    CRM_ASSERT(wl != NULL);
+
+    wl->client = request->ipc_client;
```
[the remark this answers is not in the document]
Comment (clumens) [reply daemons/attrd/attrd_sync.c:76]: Fixed both of these.

File: lib/common/ipc_attrd.c [hunk 8]
```diff
@@ -461,20 +462,24 @@ pcmk__attrd_api_update(pcmk_ipc_api_t *api, const char *node, const char *name,
     request = create_attrd_op(user_name);
     populate_update_op(request, node, name, value, dampen, set, options);
 
+    if (pcmk_any_flags_set(options, pcmk__node_attr_sync_all | pcmk__node_attr_sync_local)) {
+        ms_timeout = -1;
+    }
+
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/ipc_attrd.c:480]: I think you're right - I don't think this is necessary.  Removing and retesting.

**Comment by clumens** [comment]
I'm going to close this one and open a new PR that addresses only local sync points.

**Pull request closed without merging**